File: ConcurrencyInJava/src/concurrency/Concurrency.java
package concurrency;

/*
 *  Name: Phu Ha
 *
 * 	Project Name: CSC450- Portfolio Project
 *
 *	Project Purpose: Utilizing the concept of concurrency with the use of two threads
 *
 *	Description: There are four classes. One that include the main method to execute the program
 *		Two classes that implement the interface Runnable to be used as threads. One class is the 
 *		Counter class that will be served as the data used by the two threads.
 *
 *	Algorithm Used: N/A
 *
 *	Program Inputs: N/A
 *
 *	Program Outputs: Count from 0 to 20, then count from 20 back to 0
 *
 *	Program Limitations: N/A
 *
 *	Program Errors: N/A
 *
 *  Created on: Mar 6, 2024
 *      Author: phuth
 */

// class with the main method to execute the program
public class Concurrency {
	public static void main(String[] args) {
		
		// take note of the start time of the program
		long start = System.nanoTime();
		
		// create a counter object
		Counter counter = new Counter();
		// create the two threads that will use the counter object
		Thread thread1 = new Thread(new CountToTwenty(counter));
		Thread thread2 = new Thread(new CountToZero(counter));
		// start thread one
		thread1.start();
		// join thread one to make sure the program waits for it to finish before continuing
		try {
			thread1.join();
		} catch (InterruptedException e) {}
		// start and join thread two
		thread2.start();
		try {
			thread2.join();
		} catch(InterruptedException e) {}
		
		// take note of the end time of the program
		long finish = System.nanoTime();

		// calculate the elapse time
		long timeElapsed = finish - start;
		
		System.out.println("Time the program took to run: " + timeElapsed/1000 + " microseconds");
	}
}

// Counter class, serve as the shared data for the two threads
class Counter {
	// use volatile here to be more thread safe, will notify other threads if the value changes
	private volatile int count;
	
	public Counter() {
		count = 0;
	}
	// The following methods include the synchronized keyword to be thread safe
	// method to increment count by 1
	public synchronized void increment() {count++;}
	// method to decrement count by 1
	public synchronized void decrement() {count--;}
	
	// return count, doesn't have to be synchronized as count already has the keyword volatile
	public int getCount() {return count;}
}
	
// Thread 1, This thread will increase count in Counter class to 20
class CountToTwenty implements Runnable {
	private Counter counter;
	
	public CountToTwenty(Counter counter) {
		this.counter = counter;
	}
	
	@Override
	public void run() {
		while (counter.getCount() < 20) {
			System.out.println(counter.getCount());
			counter.increment();
		}
	}
}
	
// Thread 2, decrease count in Counter class to zero
class CountToZero implements Runnable {
	private Counter counter;
	
	public CountToZero(Counter counter) {
		this.counter = counter;
	}
	
	@Override
	public void run() {
		while(counter.getCount() >= 0) {
			System.out.println(counter.getCount());
			counter.decrement();
		}
	}
}

File: ConcurrencyInCpp/Concurrency.cpp
/*
 *  Name: Phu Ha
 *
 * 	Project Name: CSC450-Project Milestone
 *
 *	Project Purpose: To understand the concept of Concurrency in C++
 *
 *	Algorithm Used: N/A
 *
 *	Program Inputs: N/A
 *
 *	Program Outputs: Count from 0 to 20 then back to 0
 *
 *	Program Limitations: N/A
 *
 *	Program Errors: N/A
 *
 *  Created on: March 3, 2024
 *      Author: Phu Ha
 */

#include<iostream>
#include<thread>
#include<mutex>
#include<condition_variable>
#include<chrono>

using namespace std::chrono;

std::mutex mtx; // create the mutex variable
std::condition_variable cv; // this variable allows threads to communicate

int shared_counter = 0; // this variable will be use by both threads

bool finish_counting = false; // condition to see if thread one is done running

// function for first thread
void countToTwenty()
{
	std::lock_guard<std::mutex> lock(mtx); // use lock_guard to lock shared resources
	// count to 20, print each value
	while (shared_counter < 20) {
		std::cout << shared_counter++ << std::endl;
	}
	finish_counting = true; // change the condition to true, thread one is finished
	cv.notify_all(); // notify other thread
}

// function for second thread
void countToZero()
{
	// we're using a unique_lock here instead of lock_guard because
	// condition_variable works with unqiue_lock
	std::unique_lock<std::mutex> lock(mtx);

	// this asks thread this thread to wait for until finish_counting is true
	cv.wait(lock, [] {return finish_counting; });

	// count to zero and print each value
	while (shared_counter >= 0) {
		std::cout << shared_counter-- << std::endl;
	}
}

int main()
{
	// record starting time of program
	auto start = high_resolution_clock::now();

	// create thread 1 and 2
	std::thread t1(countToTwenty);
	std::thread t2(countToZero);

	// wait for both thread to finish
	t1.join();
	t2.join();

	// record ending time of program
	auto stop = high_resolution_clock::now();

	// calculate time of program
	auto duration = duration_cast<microseconds>(stop - start);

	// print out the time program need to take
	std::cout << "Time taken by function: "
	         << duration.count() << " microseconds" << std::endl;

	return 0;
}


